### ISSUE TITLE
Field group functionality in core

### ISSUE THREAD

**Issue by sirkitree** [backdrop/backdrop-issues#647]
**Update 2023:**

It has been decided that the field group module provides more options for field groups than we wish to include/support in core, so rather than move the entire module into core, we are planning to revamp the fieldgroup API and move certain types of field groups into core (groupings that already exist in core) and to leave others for the contrib module. The new fieldgroup API will become part of core's field + field_ui modules. 

Group types planed for core:
* Fieldset
* Details
* Div
* Vertical Tabs

Group types that should remain in contrib:
- HTML element
- HTML5
- Horizontal tabs
- Multipage
- Accordion

The Field Group contrib module would simultaneously release a new minor version (that depends on this version of Backdrop core) where those group types are removed from the module, and all remaining types are updated to use the core field group API (which will be similar to what's there already). The current group API will also be removed from the contrib module. 

Requirements for getting new fieldgroup contrib module in core:
- [ ] Integrate Fieldgroup API into field/field_ui modules
- [ ] Create group type for fieldset
- [ ] Create group type for details
- [ ] Create group type for div
- [ ] Create group type for vertical tabs

---

**Original Issue:** It would be pretty outstanding if we have field_group in core so that forms could be arranged a lot nicer with things like vertical tabs and such through the UI.

Requirements for getting fieldgroup contrib module in core:
- [x] Place the module into the core/modules directory
- [x] Move theme functions into field_group.theme.inc file
- [x] Remove all the field group help text and fieldset
- [x] Field Group Configure and Delete operations should use dropbutton
- [x] Field Group Configure form should use menu callback (move from an expanding section to a new page -- same as field settings).
- [x] Remove the `t()` function from all tests
- [x] Move display tests into `field_group.display.test`
- [x] Standardize on `Field group` for the UI and documentation (as opposed to `Fieldgroup` in some places).
- [x] Adjust very long lines to be under the 80 character limit for coding standards
- [x] Move display suite code to `ds` module: see https://github.com/backdrop-contrib/ds/issues/11  
- [x] Add description text for the `div` settings (what are "effect", and "speed"?)
- [x] Combine the options for `html_element` and `html5` and provide an upgrade path
- [ ] Replace the `new stdClass();` in `_field_group_get_html_classes()`.

User experience cleanup options:
- [x] N/A ~Remove the whole 'Fieldgroups' fieldset and help text~ (from previous iteration of module)
- [x] Remove the whole 'Fieldgroups' vertical tab (including the ability to clone from another display)

New features (options):
- Make field group containers collapsible
- Add a "clone" operation for a field group on the currently visible display
-  Instead of having a separate 'Add new group' row below 'Add new field', just have 'Group' as an option under 'Select a field type', and the different group types (e.g. fieldset, vertical tab, etc.) as 'Widgets'.
- (is this a good idea?) remove the ability to change the group type from the overview page (why?)
---

Related issues:
* [Switch to an "Add field" and "Add field group" action link](https://github.com/backdrop/backdrop-issues/issues/4898) (same as D8/D9):
* [Separate "Manage fields" and "Manage form"](https://github.com/backdrop/backdrop-issues/issues/5219) (field group would not appear on "Manage Fields")

Issue advocate: @jenlampton

**Comment by quicksketch** [comment]
I agree, I think this would be a useful addition. I think putting it in core is about the only way to handle it cleanly. Considering the options available today (Multifield, Fieldgroup, etc.) none of them work well because they have to make core work in ways it wasn't intended. Though if we change the way core works, then it may be possible to make this work more cleanly.

This may be a 2.x issue; I'm not sure how much API change would be required (if any) to implement this. Marking 1.x until we explore the options.

**Comment by klonos** [comment]
Yes please!!

**Comment by mikemccaffrey** [comment]
Wait, field groups have not been ported yet or merged into core? We should definitely get this done asap.

@klonos Can we keep this ticket limited to the functionality provided by Field Group, which just concentrates on organizing field, unlike Field Collection and Mulitfield, which are actually a type of field with their own data architecture?

**Comment by klonos** [comment]
> @klonos

?? I think you probably meant to mention @quicksketch but yeah, I agree. ...unless Nate wants to tackle this in a single module in a "unified" way that binds all features together.

**Comment by mikemccaffrey** [comment]
@klonos I was just apologizing to you, since I was undoing the change that you made in adding Field Collection and Mulitfield to the issue title!

**Comment by MrHaroldA** [comment]
This is a real blocker for my Backdrop installs too. Not the field collection, or multi field modules, but the possibility to add vertical tabs, horizontal tabs, fieldsets, etc through the field ui.

The Field Group module is ridden with CTools and Features specific code though.

**Comment by biolithic** [comment]
@MrHaroldA would you prefer a new module reusing the ideas of Field Group
or just a straight port replacing Ctools with our own functions?

**Comment by MrHaroldA** [comment]
Field group is big. Probably way too big. Also the UI is a bit weird when creating tabs; you have to create a tab container first.

Let's start small and slim, and gradually add features if there's a reasonable demand for it.

Adding fieldset support for example will give you a lot to work with; you could even theme them as horizontal or vertical tabs.

**Comment by serundeputy** [comment]
moving to 1.5 feel free to move back if someone wants to tackle it.

**Comment by vonnnew** [comment]
Hi All, I'm attempting my first Backdrop build now.  I really need the Field Collection module.  It doesn't appear to have been ported yet.  Is this true?  If not, I will attempt it but it will be my first attempt to port a module.

**Comment by herbdool** [comment]
Field collection hasn't been ported. At least I don't see it in the backdrop contrib list: https://github.com/backdrop-contrib/?utf8=%E2%9C%93&query=collection

Let's keep this issue focused on field group, however, which is for different purposes. I agree with @MrHaroldA about paring down field_group but it's also useful to keep it as close as possible with the Drupal 7 version for now. As far as I can tell, ctools is only being used to provide the CRUD functionality in the UI, which presumably could be replaced with Backdrop's CMI.

**Comment by jenlampton** [comment]
@vonnnew  Field collection is going to be a doozy for your first port. I'd personally been checking out multifield instead since Field collection has some serious performance implications.  Good luck!

also, no PR for this yet, pushing to 1.6

**Comment by laryn** [comment]
Found my way here on a search. To update RE the mentioned contrib modules:

- Field Collection: Not ported to my knowledge
- Multifield: [Ported](https://github.com/backdrop-contrib/multifield)
- Field Group: [Ported](https://github.com/backdrop-contrib/field_group)

**Comment by klonos** [comment]
Since 2015 (when this issue was first filed), there's a new player that's been gaining momentum: https://www.drupal.org/project/paragraphs

> Paragraphs is the new way of content creation!
It allows you — Site Builders — to make things cleaner so that you can give more editing power to your end-users.

> Instead of putting all their content in one WYSIWYG body field including images and videos, end-users can now choose on-the-fly between pre-defined Paragraph Types independent from one another. Paragraph Types can be anything you want from a simple text block or image to a complex and configurable slideshow.
>...

In fact, there's a note in the https://www.drupal.org/project/field_collection project page that says:

> [Paragraphs](https://www.drupal.org/project/paragraphs) is likely to replace field collection for Drupal 8. Field collection is on its way to being deprecated. It is recommended to use [paragraphs](https://www.drupal.org/project/paragraphs) instead of field collection for Drupal 8 projects.

I agree with @mikemccaffrey that we should keep this issue focused in getting the field group functionality in core (a good candidate for 1.9 perhaps?), but I think that we should seriously discuss about the concept of paragraphs and getting that into core too. It would be a killer-feature to have for 2.0 I think.

**Comment by klonos** [comment]
I have recently created #3766 to track progress on what's cookin' for D9 ...it seems that https://www.drupal.org/project/field_union is where things will happen.

**Comment by BWPanda** [comment]
I just installed the [Field Group](https://github.com/backdrop-contrib/field_group) module on my dev site to see how it works and get an idea of how it'd ideally work in core. Here's how it currently looks:

![Screen Shot 2020-08-16 at 13 21 50-fullpage](https://user-images.githubusercontent.com/2385329/90325809-bb2b1100-dfc3-11ea-982d-221fb1d4738b.jpg)

My thoughts:

- Instead of having a seperate 'Add new group' row below 'Add new field', just have 'Group' as an option under 'Select a field type', and the different group types (e.g. fieldset, vertical tab, etc.) as 'Widgets'.

- Remove the whole 'Fieldgroups' fieldset (I don't see the need for that much help text, and I've personally never used the Clone feature).

- Make a field group look and act more like the other fields by moving all configuration to a seperate page (i.e. remove the ability to change the group type from this overview page, and move Configuration from an expanding section to a new page (same as field settings)).

Based on my suggestions, here's how it could look:

![Screen Shot 2020-08-16 at 13 32 58-fullpage](https://user-images.githubusercontent.com/2385329/90325921-1f020980-dfc5-11ea-9af9-133808231f6c.jpg)

**Comment by herbdool** [comment]
I've usually used the clone feature. If there's a way to preserve would be good.

**Comment by jenlampton** [comment]
This looks great. Can we add a clone link into operations for the field
grpups?

**Comment by klonos** [comment]
> Can we add a clone link into operations for the field grpups?

Yes please! Would love that ❤️ 

> - Instead of having a seperate 'Add new group' row below 'Add new field', just have 'Group' as an option under 'Select a field type', and the different group types (e.g. fieldset, vertical tab, etc.) as 'Widgets'.

I'm not opposed to this, but wanted to mention that back in 2015 @wesruv has provided some mockups in https://github.com/backdrop/backdrop-issues/issues/779#issuecomment-141707916 and https://github.com/backdrop/backdrop-issues/issues/779#issuecomment-163844301 of how things could work if we switched to an "Add field" button, same as D8/D9:

![image](https://user-images.githubusercontent.com/2423362/90553076-74dcda00-e19c-11ea-9b0a-ab52a1bc7ef3.png)

UX studies by our Drupal brethren have shown that it is a much better workflow, and confuses people less than the current UI (and I prefer it too TBH, because it brings consistency in our UI patterns). In D8/9 they've separated the UI that allows adding and reordering fields into a dedicated "Manage fields" and a dedicated "Manage form display" form. So the "+ Add field" and a "+ Add field group" buttons there are in separate places. We might need to move to a UI that has the 2 "+ Add field" and a "+ Add field group" buttons side-by-side at the top-left. If you put it all together, and include @jenlampton's suggestion in @BWPanda's mockups, it would like this:

![Screen Shot 2020-08-18 at 10 24 11 pm](https://user-images.githubusercontent.com/2423362/90556526-c5a30180-e1a1-11ea-8095-0ef5eb5b0241.png)

> Remove the whole 'Fieldgroups' fieldset ...

👍 

> Make a field group look and act more like the other fields by moving all configuration to a seperate page...

Also 👍 ...UI consistency please. I would suggest that we eventually switch to using dialogs at some point, same as we've done in text formats UI in #1032, which I find to be a big UX++, as the user remains within context of the task they set off to do.

**Comment by jenlampton** [comment]
I don't know about this.. I find a LOT Of value in being able to see the fields that have already been added while adding a new one. I would hate to loose that behind a modal, or on a previous page. (This is particularly important on content types with lots of similar fields)

**Comment by klonos** [comment]
>I find a LOT Of value in being able to see the fields that have already been added while adding a new one. I would hate to loose that behind a modal, or on a previous page.

Dialogs work fine that way when adding blocks to a layout. Just saying.

>I also really like the ability to put the new field right where you want it (between other fields) when it is added, rather than needing to return at the end to order them.

Again, block placement works that way: new blocks are added to the bottom of any existing blocks in the same region; the the user needs to drag them to the desired place. But perhaps we could find a way that this works the same as the +/- buttons in option elements:

![Screen Shot 2020-08-19 at 1 09 01 am](https://user-images.githubusercontent.com/2423362/90570602-b7acab00-e1b8-11ea-87a7-1c7589809b75.png)

**Comment by jenlampton** [comment]
> Dialogs work fine that way when adding blocks to a layout.

Sure, it's "fine". But it's not as good as what we have today.

> Again, block placement works that way: new blocks are added to the bottom of any existing blocks in the same region;

Yes, and this is also a pain point that we have learned to deal with. It's all "fine", but I prefer the better UI we have today and would hate to loose it for the sake of consistency.

**Comment by klonos** [comment]
>Can we move these thoughts to another issue? I don't want a UI rewrite to block getting Field groups in core!

Sure can 👍

**Comment by olafgrabienski** [comment]
> I don't want a UI rewrite to block getting Field groups in core

Agreed! So, back to @BWPanda's suggestion? Don't see it as UI rewrite but contrib integration:

> Instead of having a seperate 'Add new group' row below 'Add new field', just have 'Group' as an option under 'Select a field type', and the different group types (e.g. fieldset, vertical tab, etc.) as 'Widgets'.

I like the idea to treat fields and groups in the same Ui. Unfortunately, some labels and descriptions don't cover the group use case, e.g. the following bold words:

- Add new **field**
  (Could be "Add new field or group". Or "Add new element"?)
- Type of **data to store**
- Form element to edit **the data**

**Comment by laryn** [comment]
Definitely in favor of adding this functionality in a more streamlined way. (I like "Add new element")

I will note that the Clone functionality at the bottom seems different than simply adding a clone option in the operations dropdown -- I've never used it, but it looks like it clones field groups from a different view mode into the one that you're looking at, not simply cloning a group from the current form to the current form.

**Comment by philsward** [comment]
FG works fine for minimal stuff. Not so great for more complex stuff. Wish FG was easier to deal with on bigger entities.

![image](https://user-images.githubusercontent.com/3436110/105805356-6ffc2d80-5f67-11eb-9478-de90ed145d4f.png)

**Comment by klonos** [comment]
Oh wow! 😅

**Comment by stpaultim** [comment]
We are approaching that final month before code freeze for version 1.21 and this is the issue that got the most votes in our priority survey last March that has not yet gotten into core (in fact, it was highest on the list at the time of the survey).  

In our recent prioritization discussions, this issue was rated "Medium" difficulty and "Medium" impact. This could be the flagship new feature of 1.21. Maybe??

My understanding from reviewing this discussion is the following (this is my attempt to summarize the discussion so far):

1. This issue is NOT about multi-field or field collections. It IS about the ability to organize fields into groups "so that forms could be arranged a lot nicer with things like vertical tabs and such through the UI" or " possibility to add vertical tabs, horizontal tabs, fieldsets, etc through the field ui."
2. There seems to be a consensus (so far) that the goal would be a paired down (simplified) version of the Field Group module. It is my understanding, that we're looking to add this functionality, but do not plan to move the existing Field Group module directly into core. However, I think the goal is to try to reproduce the functionality as much as possible. 
3. @BWPanda puts forward this specific suggestion for UI in this comment: https://github.com/backdrop/backdrop-issues/issues/647#issuecomment-674474016.

![Screen Shot 2020-08-16 at 13 32 58-fullpage](https://user-images.githubusercontent.com/2385329/90325921-1f020980-dfc5-11ea-9af9-133808231f6c.jpg)

4. I'm reading interest in a "clone" function - possibly in the operations button.
5. @klonos made some suggestions that might be best treated as follow-up issues. 
6. @philsward pointed out limitations of basic field group functionality when there are LOTS of fields involved. But, it's not clear to me that this was intended nor needs to slow this issue down?

NEXT STEPS:

In my opinion, this issue needs two things:

1) An advocate 

AND/OR

2) Someone with the technical skills to create a first draft PR (to generate momentum). 

QUESTIONS FOR DISCUSSION:

Would it be realistic to get this committed before feature freeze?
Is anyone interested in taking some leadership on this issue?

**Comment by philsward** [comment]
To clarify my position re: @stpaultim 

I don't think my concerns need to hinder any core inclusion.

Most use cases won't ever run into the bad UX of 30+ fields on an entity. Having a way to group is the first step to a better overall UX and eventually a better UI.

**Comment by klonos** [comment]
Thanks @stpaultim for taking the time to summarize this 🙏🏼 

> Someone with the technical skills to create a first draft PR (to generate momentum).

I definitely think that most if not all suggestions here can be follow-ups (we can add them to #779) - what we need is indeed something to start iterating on. Baby steps.

We already have other issues for moving the 'Add new/existing field' functionality to a new form via an action link: #4898. We can take discussions re that there.

Can someone please create follow-ups for:
- Adding a "clone" operation
- expanding/collapsing field groups to save vertical space

**Comment by stpaultim** [comment]
@herbdool - I've noticed quite a bit of activity on Field Group module the last week and a new release. Does that impact this issue at all?

It was my understanding that the plan was to recreate Field Group functionality in core, not necessarily include the field group module in core. Has or should that plan change? Is it possible?

**Comment by herbdool** [comment]
@stpaultim not sure. Depends on how Field Group gets put into core.

I personally think it's best to make some of the UI improvements suggested above in the contrib module so they can be confirmed how they work. Such as change the operations to drop down on Form (though on Display the default is not a drop-down operations so it needs a different solution). Some of the ideas above need more thinking through so why not create a PR for contrib and see how it sticks.

I don't find Field Group difficult to use. Some of the types I haven't used before but I just tested them and found them straightfoward: vertical tabs, horizontal tabs, etc. I've used the cloning quite a bit: clone all field groups from form to display and vice-versa.

Trying to recreate Field Group is the most difficult path to take. Plus there's no good reason to do it in isolation. Even if the functionality is recreated, Field Group in contrib should end up using core's API and be redirected to provide the extras. Field Group right now has a robust enough hook system that that can be done.

If someone wants to step forward to work on this, I recommend first creating PRs for contrib to improve the UI there. And that can shifted to core later.

**Comment by ericfoy** [comment]
As pretty much a late-comer and very much an outsider to this discussion, I thought I might offer y'all my input as "a guy who's been using Drupal (and now a die-hard Backdrop enthusiast) for a couple decades (since D5)"... Sometimes a clear voice from outside can bring the detached objectivity that helps find clarity.

One thing I had to "figure out" along the way was that there's a distinction between the Field Definition UI and the Node Display UI — and that the Field Definition UI is apparently also the Edit Form UI at the same time; whereas the Node Display UI was something different.

From a historical standpoint, this makes perfect sense, since we were originally dealing with a primarily hard-coded data structure and fieldset (until the CCK came on the scene). In the beginning, the Edit Form was roughly analogous to the Display of every node. Then through theming, etc., the actual Display (the Presentation) could be altered. This was nice.

Now we have very clearly delineated and differentiated Display Modes, which is awesome. But the same "Presentation" differentiation has not developed on the Edit Form side of the equation. We are instead still (IMHO) plagued with the direct association of the data structure with the Edit Form Presentation. This, I believe, is what leads to discussions about whether the fieldgroup should be added inside the same UI steps as the field (to which I would mildly object, because they're not the same).

With the utmost respect to those who have actually done the work, and contributed to the development of this fine product, I offer my "outsider's opinion" on the issue (this is obviously a long-term discussion, and points to BD2 or beyond):

The UI that I envision at admin/structure/types/manage/[content-type] would have the Tabs: 

- Configure | Manage Fields | Manage Displays | Manage Forms

Fieldgroups therefore clearly would belong not under the "Manage Fields" tab, but under the "Manage Forms" tab. Furthermore, it is quickly seen that the various widgets would no longer (necessarily) be owned by the fields themselves, but rather by the "Form Mode" as defined under the "Manage Forms" tab. Perhaps the widgets should still be defined for the "Default Form" via the Add Field dialogs as they now are, and then alternatively be overridden for any additional Form Modes under the Manage Forms tab. This would rightly include Fieldgroups.

I am especially addressing the clear difference between Field Groups and Field Collections/Multifield, which I am sure everyone already recognizes, but if I may, I will stress this point, which I have not heard explicitly expressed: "Field Groups" is a Presentation thing, whereas "Field Collections/Multifield" are a data structure thing. And it is for this reason that I would humbly object to the further entrenchment of the melding of these two functions.

Realizing that this idea represents a significant departure from how it's always been done, and that a lot of significant recoding would be required, I nonetheless offer the following "really cool" results that might follow:

- The concept of "Editable Fields" could be greatly expanded... "edit fieldgroup," field groups with edit priveleges, etc.
- Context-sensitive edit forms of various layouts.
- Templates for node-edit forms, ready-made from the UI
- Differing widgets for differing edit contexts.

Being an outsider, It's really easy for me to just MAKE THIS STUFF UP, like it's no big deal, right? Nonetheless, being a bit of an OCD data and presentation idealist, I felt the need to put these ideas forth for your consideration. Put them on the back burner if you will.

Sincerely, —Eric

**Comment by MrHaroldA** [comment]
> Fieldgroups therefore clearly would belong not under the "Manage Fields" tab, but under the "Manage Forms" tab.

I'd say they should be on the "Manage Displays" tab as well ...

**Comment by bugfolder** [comment]
I very much agree with ^^^. (It took me an embarassingly long time in Drupal to discover that rearranging the data structure was also rearranging the edit form.) And the ability to create multiple versions of the edit form would be GREAT. I've got a use case now where ordinary people edit a subset of node fields on a bespoke edit form, while admin users can access more of the fields on a different edit form; being able to set up something like that from the UI (rather than hacking it in code, as now) would be great.

**Comment by laryn** [comment]
These issues may be related (at least tangentially):

 - https://github.com/backdrop/backdrop-issues/issues/3072
 - https://github.com/backdrop/backdrop-issues/issues/2608#issuecomment-291273853

**Comment by stpaultim** [comment]
We talked about this a bit in the dev meeting yesterday - https://youtu.be/KLihQQK7f-M [22:35]. 

There was some discussion about whether or not this might be possible for next release. My sense is that we don't have enough consensus of which direction this is going for anything to happen too quickly. There were some hints that folks are skeptical of putting the field group module directly into core, for these reasons:

1) It doesn't seem like the kind of functionality that you need to or want to turn off or disable. 
2) It might be better to include this functionality in a general reworking of the field ui.

I don't know that anyone was firmly behind or against any particular approach yet, which is another reason that it might be a while before we are ready to tackle this. I believe that for this issue to progress, we need an advocate for this issue with a better understanding of the technical implications, a willingness to push a specific solution, and the ability to work on a PR. I'm not that person.

**Comment by herbdool** [comment]
@ericfoy sounds like how Drupal 8/9 divides the functionality. Or is because of the experimental module for form layout in Drupal. I can't recall but that's what comes to mind.

**Comment by herbdool** [comment]
@stpaultim the reworking of the UI might be good idea, though I wonder if there's an MVP that could be done in stages.

Field Group wouldn't need to be a separate module but could still copy code out of the module much like was done with Token. However, it would still need to be a way to transition from the contrib module and also have plugins for expanding the group types. So those not included in core can still be used.

**Comment by jenlampton** [comment]
I'd like to advocate for this issue in Backdrop 1.23.0.

**Comment by jenlampton** [comment]
Okay, I've posted a first draft PR that accomplishes a few of the recommended changes for field group, there's plenty of room for more improvement, but I think this is a good start!

**Comment by laryn** [comment]
@jenlampton Did I understand correctly that you addressed all of @herbdool feedback in the PR? If so let's remove "Needs work" here so it's clear it just needs testing.

**Comment by herbdool** [comment]
I don't think it's there yet. There's leftovers from the previous way of configuring and I don't think it's quite working yet.

I also wonder if it might be easier to merge in this module if we first make some of the changes in the contrib module and then merge. As with entityreference it is a big thing to review. Separating out merging the contrib module from improvements required before it can be merged would make it much easier. I think converting the links to dropbuttons, and adding a clone option to the dropbuttons could be done in contrib first.

**Comment by herbdool** [comment]
Pretty clear this won't be ready in time for 1.23.0 so bumping.

**Comment by jenlampton** [comment]
> I also wonder if it might be easier to merge in this module if we first make some of the changes in the contrib module and then merge.

We'd need to remove the "clone" feature from the contrib module in order to do that (which, I expect might upset some people).

That said, I'd be happy to re-purpose some of this PR to apply it directly to the contrib module first if that's valuable.

**Comment by herbdool** [comment]
We can remove that one bit in the pr to core and leave it in contrib. Though eventually if it's included in core and force people to disable it in contrib then it won't make any difference.

My pr already copied your PR plus made a few more fixes.

There's still some funny UI stuff that needs to be fixed.

**Comment by stpaultim** [comment]
@herbdool - It's my understanding that this issue has an advocate and is therefor automatically eligible for the next minor milestone. @jenlampton are you still advocating for this issue?

**Comment by stpaultim** [comment]
There was discussion during the last DEV meeting about seeing if we can make progress on this. This could be the highlight of the next minor release. If I understood correctly, it primarily needs a little testing and review. Anyone?

**Comment by olafgrabienski** [comment]
> If I understood correctly, it primarily needs a little testing and review. Anyone?

I'd love to test the PR, but don't know what exactly to look at. @jenlampton and/or @herbdool can you give a summary of the current state and indications for testing? The Tugboat sandbox is expired, btw.

**Comment by klonos** [comment]
@jenlampton this is indeed a huge change to be reviewing. Perhaps if we had a way to compare only commits after the contrib module was added (almost verbatim I'm assuming) as a starting point, then it'd be easier. Anyway, I don't know if there were other commits before the initial one, but I started reviewing one commit at a time as they are listed in https://github.com/backdrop/backdrop/pull/4076/commits

- I liked the removal of `field_group_format_settings_label_validate()` and replacing that with making the label field always required 👍🏼 ...it's not 100% match with the previous behavior in the contrib module, but it is much simplified. The only caveat would be when re-saving forms where the label was left blank (because the contrib module allowed that in certain scenarios). Not a biggie though, and perhaps we can get an update hook that is adding some arbitrary labels to field groups with missing labels(?).
- Re making the "Separate class names with spaces. ..." description a reusable constant that contrib could reuse, I'd frankly like to get rid of this entire concept. I don't think that it's good UX to be telling people to "leave this field blank to do x" or "do not add a leading forward slash" and if they don't adhere to the help text guidelines, then be throwing errors after form submission 🙄 ...I mean, really? Can't we just remove the leading slash in the form validation? ...so similarly with CSS classes/IDs, I would like us to allow either commas or spaces, and then transparently remove commas in the form/field validation (the same way we `trim()` values etc.).
- I like the idea of making the ID field a `'#type' => 'machine_name'` ...it forbids capital laters though, which although not common, they are still valid CSS. Perhaps add a custom `replace_pattern` of `[^a-zA-Z0-9_-]+` ...in fact. since CSS classes/IDs should not start with a number, the regex should be `^[a-zA-Z]{1}[a-zA-Z0-9_-]+$`, which ensures that the first character of the class/ID must be a non-numeric character. If we really really want to be accurate, we should tweak that pattern to match the spec in https://www.w3.org/TR/CSS2/syndata.html#characters:
  > - All CSS syntax is case-insensitive within the ASCII range (i.e., [a-z] and [A-Z] are equivalent), except for parts that are not under the control of CSS. For example, the case-sensitivity of values of the HTML attributes "id" and "class", of font names, and of URIs lies outside the scope of this specification. Note in particular that element names are case-insensitive in HTML, but case-sensitive in XML.
  > - In CSS, identifiers (including element names, classes, and IDs in [selectors](https://www.w3.org/TR/CSS2/selector.html)) can contain only the characters [a-zA-Z0-9] and ISO 10646 characters U+00A0 and higher, plus the hyphen (-) and the underscore (_); they cannot start with a digit, two hyphens, or a hyphen followed by a digit. Identifiers can also contain escaped characters and any ISO 10646 character as a numeric code (see next item). For instance, the identifier "B&W?" may be written as "B\&W\?" or "B\26 W\3F".

  ...the regex should most likely go into a reusable helper function BTW.
- Speaking of which, I don't think that `#exists` is a valid property for `'#type' => 'machine_name'` - it needs to be an `exists` key (note: no `#`) inside a `'#machine_name'` property array. If it is valid syntax, then we need to update our FAPI documentation. See: https://docs.backdropcms.org/form_api#machine_name
- I would much prefer it if we had separate FAPI elements for CSS classes/IDs, since both core and contrib is using those in settings forms.
- Not sure about the removal of the help text - there are some things that aren't straight-forward, unless you have used field groups before. We could be throwing "dirty form" warnings (for example when accordion items or vertical tabs are "parentless") ...or we could be automagically adding wrapper groups (somehow?). In the meantime, perhaps use the same "collapsed fieldset help" thing that we have also done in the layouts listing page?

That's it for now - I'm sure I must have missed certain things, but once we have the main functionality in, we can iteratively improve/fix things.

**Comment by jenlampton** [comment]
> I would much prefer it if we had separate FAPI elements for CSS classes/IDs, since both core and contrib is using those in settings forms.

Same.  Could you create an issue for that? :)

**Comment by jenlampton** [comment]
> I like the idea of making the ID field a '#type' => 'machine_name' 

Now that I've read through your whole comment, I don't actually like using `machine_name` for a HTML ID attribute. Using a machine name would enforce that there is only one fieldgroup with each ID, and you might want one on each node type with the same ID, so I don't think this was a safe choice...  I was trying to avoid the additional validators, but perhaps just removing it would be sufficient. 

> Not sure about the removal of the help text - there are some things that aren't straight-forward, unless you have used field groups before. 

The concept of grouping fields is simple though, and if we can guide people through using the form correctly as they are doing it, that would be a better experience than making people read about scenarios that may not apply to them.

> We could be throwing "dirty form" warnings (for example when accordion items or vertical tabs are "parentless")

This does seem to be happening already: 
<img width="1222" alt="Screen Shot 2022-12-30 at 3 55 50 PM" src="https://user-images.githubusercontent.com/397895/210118791-f3cf9da4-c4e0-466d-bd54-50949523ea9d.png">

**Comment by jenlampton** [comment]
I filed [a new PR](https://github.com/backdrop/backdrop/pull/4289) that preserves the git history from field_group, and I'm hoping the sandbox will work on that one also 🤞

**Comment by jenlampton** [comment]
It looks like I've got some failing tests, and we could use a little CSS cleanup for Seven, I'll work on those tomorrow. 
But the sandbox is back @klonos ! https://pr4289-ntjttt4v4pproevelni7ti26smnkxk3t.tugboatqa.com/node/add/page

**Comment by jenlampton** [comment]
I've fixed the failings tests! (I am also getting what I think is a random failure on `LanguageUpgradePathTestCase`, for only one version of PHP).  I've got those running again and will start on some CSS cleanup for Seven (and I will test Basis also).

**Comment by jenlampton** [comment]
I've just noticed that field_group calls it's thingies both "Fieldgroups" (one word) and "Field groups" (two words) in both the user interface, and in code comments. I'm going to standardize on "Field groups" (but we can change that after Jan 1 if people prefer "Fieldgroups" :)

**Comment by bugfolder** [comment]
> but we can change that after Jan 1 if people prefer "Fieldgroups"

I'll put my vote in for "Field groups."

**Comment by jenlampton** [comment]
One of the tests was failing because fieldgroup allowed people to select a "label element" for a "Details" group, but the Backdrop core details element always uses a `span` for this. The rendered output always added a second '<span>' around the "label element" that was selected, resulting in `<span><span>`, where the test was expecting only one. 

My solution was to remove the label element, but I'm now thinking maybe I should have updated the test to search for the extra span, as the current solution would loose the functionality of being able to specify that tag...  I'm going to go put that back.

**Comment by quicksketch** [comment]
I haven't used Field Group module any time recently, so a lot of this module is new to me. In my own sites, I tend to form_alter() to achieve two things:

* Put items into vertical tabs (I do this for meta-related information on almost all sites)
* Put fields in fieldsets

Reading this PR I see that Field Group module does a _lot_ more than this. A lot of things make me nervous, like multipage and accordians. Do you think it would be feasible to instead put the API from Field Group module into core as part of Field module directly, and then make a "Field Group 2.x" that used the API to provide all the current functionality? Then maybe core could only worry about arranging forms and the most common features; and Field Group would maintain and provide an upgrade path for existing users of some of the more esoteric grouping options.

**Comment by olafgrabienski** [comment]
@quicksketch What's the reason for your nervousity: the number of features and options, or rather the technical architecture? I'm not able to evaluate the latter, but hope it can be improved by and by, if necessary. Re differentiating between 'most common' and 'more esoteric' features / options: While some use vertical tabs and fieldsets, others will want to use horizontal tabs and DIVs, and others even more. Which ones are common enough to be integrated in core?

Generally, as a site architect, I want to stay flexible and have all reasonable *Field Group* features in one place. And I'd like to avoid putting too less in core, forcing too many people to add the contrib module anyways.

**Comment by klonos** [comment]
Yeah, I'm also with @olafgrabienski on this @quicksketch. Field Group has a HUGE user base in Drupal, and it is a great low-code/no-code site builder tool (DX+) that helps bring sanity to content type forms (content editor UX+). It reminds me of the CCK and Views cases, where really popular modules got into core. We didn't include only portions or the API or only Views without the Views UI - we added these almost as they were.

Besides an indication of popularity, the mere fact that ~300k sites are using the module should be providing us enough confidence for its stability. Health-wise, there are currently 180 open bug reports in the d.org queue for the module, but only 86 of them for the 7.x version, and only 1 critical (which concerns an upgrade to a very dated version, from v1.3 to v1.4 which was released in 2014, whereas current is v1.8 released in October).

In Backdrop land, although they don't do the same thing, Field Group and Paragraphs are equally popular, with ~500 installations; there are only a handful of open issues in its queue; one of our core maintainers is already the maintainer of the contrib module, and it seems that we have @jenlampton and myself backing this (among other very active contributors in the community). These should all be things to factor in and provide some more reassurance and confidence. Right?

**Comment by olafgrabienski** [comment]
> ... the sandbox is back

I've had a look at the sandbox site yesterday and today. It worked generally fine, but today there were a few issues. (I didn't see the errors yesterday, the issues seem to have been introduced inbetween.)

- After enabling, today I saw the following message:
  `ParseError: syntax error, unexpected ';', expecting ',' or ')' in _theme_process_registry() (line 158 of /var/lib/tugboat/core/modules/field_group/field_group.theme.inc).`
- After editing a HTML element group, today I saw this message:
  `Notice: Undefined index: formatter in field_group_configure_form_submit() (line 787 of /var/lib/tugboat/core/modules/field_group/field_group.field_ui.inc).`
- Message while editing a Details group today:
  `Notice: Undefined index: label_element in field_group_field_group_format_settings() (line 714 of /var/lib/tugboat/core/modules/field_group/field_group.module).`

**Comment by olafgrabienski** [comment]
When testing Field group, two questions came to my mind:

1. Is Field Group supposed to be enabled by default? (In the sandbox, I had to enable the module.)

2. Can we make the wording of the multi group types more consistent? It's just a small change, remove the term "item" from "Horizontal tab item".

- Vertical tabs group / Vertical tab
- Horizontal tabs group / Horizontal tab ~~item~~
- Accordion group / Accordion item

**Comment by herbdool** [comment]
@quicksketch it's divided fairly cleanly between API and the group types so it can be done relatively easily. Just need to think of how to transition users with the upgrade so they use the right contrib module. I imagine a new contrib module would be needed for the extras and current contrib would be disabled upon upgrade.

**Comment by herbdool** [comment]
@quicksketch just reread your comment and understand better. If we put the API into Field module then we don't have to create a new contrib module. That'll be easier. Makes it harder to preserve commit history though?

**Comment by herbdool** [comment]
I think separating the functions out should first happen in contrib so it's easier to move around.

**Comment by klonos** [comment]
> Is Field Group supposed to be enabled by default?...

If we decide to separate the UI (and implementations) out to contrib and only add the API into core, and also move the API into the Field module, then I guess that the API would always be enabled/available (the Field module is always required/enabled).

If we stick with how the module is in contrib now, then it wouldn't make sense to have it disabled once added to core. What would the point of it all then be? It's relatively the same to have the module in core and disabled (where people need to enable it to be able to use its features) and have it in contrib (where people can get it via the Project Browser, enable it, and then use it). I thought that the goal was to have it enabled OOTB + then perhaps follow up with a way to demo its use in one of the existing content types. That way people would be able to get an example use case and tweak further (rather than having to discover that this feature exists in the first place).

**Comment by jenlampton** [comment]
> Do you think it would be feasible to instead put the API from Field Group module into core as part of Field module directly, and then make a "Field Group 2.x" that used the API to provide all the current functionality? Then maybe core could only worry about arranging forms and the most common features; and Field Group would maintain and provide an upgrade path for existing users of some of the more esoteric grouping options.

I really like this idea. After doing a lot of work on the module I also learned about strange things that it does that probably shouldn't be in core (yes accordions and multipage elements, but also collapsible divs?!?). 

I was previously more concerned about the upgrade path, and wanted everyone running field_group to NOT need to install a field_group_extras module to get the same behavior after the core update.

If we, instead, moved the API, and maybe just a few group types (`fieldset`, `details`, `and vertical tab` come to mind)  into the core Field UI module, everything else could remain in field_group, and people previously using field_group would have a seamless upgrade path.

> While some use vertical tabs and fieldsets, others will want to use horizontal tabs and DIVs, and others even more

There's one pretty clear place to draw a line: Core could support all the elements we already have in core, and everything else could remain in contrib. We'd essentially be adding an API, and then implementing the API to support everything that already exists. 

> What's the reason for your nervousity: the number of features and options, or rather the technical architecture?

For me it's both. 
* Features and options: There's a lot of features that I think are edge cases that most people would never touch. But I do like the flexibility of making those things possible - hence the API. 
* Technical architecture: The field_group API could also use a minor update (callback functions please). Also, trying to get the module to meet coding standards as been rough. 

> I think separating the functions out should first happen in contrib so it's easier to move around.

I'm happy to help with that. I would also like to contribute some individual patches with things like UX improvements and coding standards cleanup so that the work from the current PR isn't lost.

@herbdool I could see field_group 2.0 simply switching from the current API to the new core API, and removing the group types now provided by core. Core could handle any configuration changes that were necessary for those group types (If there are any, but I'd prefer there weren't!) and as long as the 2.0 version has a dependency on the version of core that gets the API, the upgrade path should be smooth.

> Makes it harder to preserve commit history though?

Yeah, I can't think of how we could do that...

**Comment by quicksketch** [comment]
> Generally, as a site architect, I want to stay flexible and have all reasonable Field Group features in one place. And I'd like to avoid putting too less in core, forcing too many people to add the contrib module anyways.

I suppose I'm not in the majority here, but I've personally never used Field Group module, and if I did, it would be to only organize the edit form fields, not to modify the front-end display. I think I'm most worried about the burden of needing to support entirely new display styles like horizontal tabs on the front-end. Basis will probably need to be updated to at least minimally support every Field Group display style, as will any future core themes.

> If we put the API into Field module then we don't have to create a new contrib module. That'll be easier. Makes it harder to preserve commit history though?

That's right, what I was thinking is that Field module would adopt the API, and provide vertical tab and fieldset group types. Then Field Group module would continue to live in contrib and provide all the other group types.

But if we were to follow that plan, developers that use Field Group as a site-building tool would very likely still need to use the contrib module.

**Comment by klonos** [comment]
> ...I do like the flexibility of making those things possible - hence the API.

My only concern with this thinking is that the API is a developer tool - excluding the current implementations from the contrib module would mean that people still need the contrib module.

> I've personally never used Field Group module, ...

Yes, because as you've mentioned in one of your earlier comments, you code your way out of things required. Not everyone can do that though.

> ...and if I did, it would be to only organize the edit form fields, not to modify the front-end display.

Yes, this is a great tool for site builders to provide friendlier forms for content editors with minimal effort (no code). That's why I like it.

> fieldset, details, and vertical tab come to mind

In GovCMS, I've seen frequent use of horizontal tabs, as well as the sidebar implementation (but that's only a D8+ thing, so not in the 7.x version of the module).

**Comment by herbdool** [comment]
@jenlampton @quicksketch I think we're on the same page.

**Comment by jenlampton** [comment]
I've updated the main issue with the current plan as I understand it. Resetting labels.

**Comment by docwilmot** [comment]
Is someone working on it though?

**Comment by stpaultim** [comment]
This issue just came up in Zulip. It is currently item 4 on our [Feature Request Wishlist](https://forum.backdropcms.org/feature-requests). 

There was a lot of excitement about this issue a year ago and some of us thought it was very close to going into core. As I review the most recent comments, it seems that some reconsideration is happening (or has been decided). As I understand it, both @jenlampton, @quicksketch, and others now believe that this module MIGHT be doing more than necessary or at least more that we wish to support in core.

The current plan seems to be to move some of the API's into core and to keep the User Interface in contrib. The original issue has been updated to reflect this. 

Personally, I understand that this might be the smart course of action, but I don't remember the discussion and am concerned that this was a popular feature request, because folks wanted the UI in core not just the API. Moving the API may benefit developers, but it will not benefit site architects/builders.

I'm not necessarily opposed to this plan, but I would like clarification. While supportive of either option, I less excited about the API only idea.

**Comment by kiamlaluno** [comment]
@stpaultim Reading the issue summary, I gather the aim is to include in core some of the group types the contributed module provides, not to include only the module API in core, at least because there is the *Integrate Fieldgroup API into field/**field_ui** modules* list item.

**Comment by klonos** [comment]
Yup, what @kiamlaluno said is right @stpaultim - the decision was to add the UI for this as well, but to provide support for only some of the most common grouping elements and then leave the rest of them for the contrib module.

**Comment by stpaultim** [comment]
Thanks to @kiamlaluno and @klonos for clarification. I will edit the original issue to make that a bit clearer, but I think it was the most recent comments in issue that confused me. 

The next thing I'm looking for clarification on is what is the next step. Does this mean we're waiting on someone to start refactoring the existing contrib module to move some code into a completely new PR, while revising the old contrib module to work with the API's that will now be in core?

Is there any usable work that has been done already or are we looking for someone to start over on this task?

**Comment by kiamlaluno** [comment]
There were two PRs, but one was closed in favor of the other one, and https://github.com/backdrop/backdrop/pull/4289 was said to be rejected by core committers.  
I guess this needs to be started over, but I am not sure which approach should be followed, nor why the previous approach has been rejected.

**Comment by kiamlaluno** [comment]
(Never mind: The new plan should be the one described before *Original issue*, which in fact has a list with no checked item.)

**Comment by kiamlaluno** [comment]
I am in favor of this feature, but I do not know how to contribute. The first point (*Integrate Fieldgroup API into field/field_ui modules*) is rather broad. I would not want what that should be done while keeping the commit history from the contributed module.

**Comment by klonos** [comment]
Here's a rough plan the way I understand things:
- Decide if the module is going to be merged as is in core, or integrated into the existing field and field_ui core modules. This is important, as it will have to do with namespacing and the contrib module + tell us if we need to add the module name in [`backdrop_merged_modules()`](https://docs.backdropcms.org/api/backdrop/core%21includes%21module.inc/function/backdrop_merged_modules/1) or not.
- Merge the contrib module code into core, and start moving things around as appropriate, renaming, and removing things as per the plan outlined in the issue summary.
- Create a 1.x-2.x branch in the contrib module, and move any code that has not been included in core there. The .info file of the module in that new branch should have `dependencies[] = system (>= 1.27.0)` in it.
- Add `dependencies[] = system (< 1.27.0)` in the .info file of the contrib module in the 1.x-1.x branch.

**Comment by herbdool** [comment]
@kiamlaluno the plan is to get the contrib module in a state that is acceptable for core first. This may mean making a sub-module to move the features there, which won't be merged into core. Then when core is released, make a new version of the contrib module that is only for those who've upgraded.

So you can contribute by going to the contrib module and helping with blockers.

**Comment by stpaultim** [comment]
I asked ChatGPT to summarize this discussion. This is the unedited response. I have yet to fact-check this response. Can anyone familiar with this issue comment on the quality of this summary?

---

## **Summary of Issue #647: Field Group Functionality in Core**

### **Background & Proposal**
- The discussion began in 2015, proposing that **field grouping** (e.g., vertical tabs, fieldsets) should be included in **Backdrop CMS core** to improve the content editing experience.
- Initially, the idea was to integrate the **entire** Field Group module into core, but later discussions favored a **split approach**: 
  - Move **some** grouping types into core.
  - Keep **others** in a contrib module.
  - Introduce a **revamped API** to support these changes.

### **Final Plan**
- **Field group types to be added to core:**
  - **Fieldset**
  - **Details**
  - **Div**
  - **Vertical Tabs**
- **Field group types that remain in contrib:**
  - **HTML element**
  - **HTML5**
  - **Horizontal tabs**
  - **Multipage**
  - **Accordion**
- The new **Field Group API** will be part of core's `field` and `field_ui` modules.

### **Implementation & Roadblocks**
- **Integration Approach**: The preferred method is **integrating the API into core** while keeping some advanced features in contrib.
- **Commit History Issues**: Developers debated whether to **merge the contrib module as-is** or **split the functionality into core modules**.
- **Contrib Module Transition**: A **new major version** of the Field Group contrib module will be released to ensure compatibility with the new core API.

### **Current Status & Next Steps**
- A pull request (PR) was submitted to integrate **basic field grouping** into core, but **it was rejected by core committers**.
- The revised plan is to **first clean up the contrib module** before merging.
- Contributors are needed to:
  - Refactor the **Field Group contrib module** to separate core-ready features.
  - Transition the **API** into `field_ui`.
  - Maintain an upgrade path for sites using the contrib module.

---

This summarizes the **long-standing debate, changes in approach, and the current action plan** for bringing field grouping into Backdrop CMS core. The core team aims to **balance usability, maintainability, and flexibility**, ensuring that basic features are in core while keeping advanced ones in contrib.

Let me know if you need more details or a different focus! 🚀

**Comment by stpaultim** [comment]
In our Feb 27th dev meeting we discussed an experiment to take issues like this one that are VERY long and close the issue and reopen a new issue with a nice issue summary. 

We are trying this with this issue. Here is the new issue. Please, read through the issue summary on the new issue and edit it to add any important details that you think might be missing.

https://github.com/backdrop/backdrop-issues/issues/6865
